### PR TITLE
Solve ports conflicts with same address and different protocol

### LINKED ISSUE #711: discoveries not isolated with different protocols
**Describe the bug**
If a custom discovery lists a port with the same address as another port, but a different protocol, then Arduino-IDE should treat them as separate ports. This is how the Arduino-cli works and the 1.X IDE. 

The 2.0 IDE groups them together and only shows one device, with the protocol determined by which discovery reported the added device first. 

**To Reproduce**
 1. Edit the code for serial-discovery to return protocol="newprotocol" instead of serial and compile it with the name newprotocol-discovery. 
 2. Then configure a platform to use the newprotocol-discovery as a discovery source. 
 3. Then when you plug in a comm port, instead of see it listed twice, once with protocol 'serial' and once with protocl 'newprotocol' you will only see it once. If you unplug and replug the device, it may show up with a different protocol depending on who discovery say the reconnect first.

**Expected behavior**
In the setup above, any comm port connected to the system should be listed twice, once under 'serial' protocol, and once under 'newprotocol'. 

**Debug Lines**

For my testing, I have an Arduino device. serial-discovery sees it as:
```
 "ports": [
    {
      "address": "COM16",
      "label": "COM16",
      "protocol": "serial",
      "protocolLabel": "Serial Port (USB)",
      "properties": {
        "pid": "0x0941",
        "serialNumber": "90742540",
        "vid": "0x1781"
      }
    }
]
```

A second discovery I'm running (quarto-discovery) sees that same port as

```
 "ports": [
    {
      "address": "COM16",
      "label": "Quarto (COM16)",
      "protocol": "sam-ba",
      "protocolLabel": "sam-ba emulator",
      "properties": {
        "mi": "0x00",
        "pid": "0x0941",
        "serialNumber": "90742540",
        "upload": "1",
        "vid": "0x1781"
      }
    }
]
```
I then run arduino-ide rc2 in verbose mode and when I plug in my device, I sometimes  get:

```
root INFO Attached boards and available ports changed:
root INFO  - Attached board: Quarto [qnimble:imxrt:quarto] on COM16 Quarto (COM16)
root INFO ------------------------------------------
daemon INFO from discovery builtin:serial-discovery received message type: add, port: COM16
root WARN Port 'COM16' was already available. Known boards before override: [{"fqbn":"qnimble:imxrt:quarto","name":"Quarto","port":{"address":"COM16","protocol":"sam-ba","label":"Quarto (COM16)"}}]
```
In this case, the serial-discovery return is ignored.  Other times, I get

```
daemon INFO from discovery builtin:serial-discovery received message type: add, port: COM16
daemon INFO from discovery qnimble:quarto-discovery received message type: add, port: COM16
root WARN Port 'COM16' was already available. Known boards before override: []
```
and in this case quarto-discovery's return is ignored.

I've run the same test with arduino-cli and there I get:
```
INFO[0186] from discovery builtin:serial-discovery received message type: add, port: COM16
INFO[0186] from discovery qnimble:quarto-discovery received message type: add, port: COM16
```
where the two discoveries do not conflict and it adds COM16 twice, once with each protocol.

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Version: 2.0rc2

### PR DESCRIPTION
**Pull request by silvanocerza**
Fix ports not being stored correctly causing ports found by multiple discovery with same address but different protocol conflicting with each other and not being shown correctly in Board Selector dropdown list and Select Board dialog.

Ports separation in Tools > Ports menu is unchanged.

Fixes #711.

Board Selector dropdown
Select Board dialog
Ports menu

### CODE REVIEW

**Comment by silvanocerza** [comment]
@ben-qnimble could you test this please? 

You'll find a built artifact for each platform [here](https://github.com/arduino/arduino-ide/actions/runs/1612232354) as soon as the build process has finished.

**Comment by ben-qnimble** [comment]
Thanks @silvanocerza for the quick change! I can confirm that now ports with shared addresses but different protocols are now listed separately and no not fight each other.

I don't know if this is a separate issue or related to your fix, but if I select, say, COM5 with protocol 'something' then when I upload, the device disconnects and both COM5 ports disappear (as expected) and  when the device reconnects, both COM5 ports show up again, but the selected port is now COM5 protocol 'serial' and not 'something'. What is means is that if you click upload, make some change and click upload again the upload will potentially fail is it is connected to a different port than was  originally selected. 

Thanks,

**Comment by silvanocerza** [comment]
> I don't know if this is a separate issue or related to your fix, but if I select, say, COM5 with protocol 'something' then when I upload, the device disconnects and both COM5 ports disappear (as expected) and when the device reconnects, both COM5 ports show up again, but the selected port is now COM5 protocol 'serial' and not 'something'. What is means is that if you click upload, make some change and click upload again the upload will potentially fail is it is connected to a different port than was originally selected.

This is probably caused by the [1200bps touch](https://arduino.github.io/arduino-cli/0.20/platform-specification/#1200-bps-bootloader-reset), probably I should handle reselection of the correct port too. 

Thanks.

**Comment by per1234** [comment]
>the selected port is now COM5 protocol 'serial' and not 'something'.

There are several issues related to the IDE's handling of the port selection after the port is lost (as is normal and expected during an upload to a native USB board). That includes this issue as well as those linked from it: https://github.com/arduino/arduino-ide/issues/710

So I don't think this is specific to a port being discovered by multiple discoveries.

**Comment by silvanocerza** [comment]
I confirm what @per1234 is saying, I investigated a bit and it's not an easy fix reselecting the correct board, I risk causing even more conflicts. :confused:

**Comment by ben-qnimble** [comment]
I did a pull request (https://github.com/arduino/arduino-ide/pull/765) into this branch to add more filtering on reconnecting ports so on reconnect, the IDE don't connect to a port whose protocol does not match the protocol of the previously selected port.

**Comment by ben-qnimble** [comment]
Hi @silvanocerza I really appreciate your quick patch for this issue and I wanted to see if there was anything I could do to help move this patch forward into the main branch. 

Thanks!

**Comment by silvanocerza** [comment]
Hey @ben-qnimble, am waiting for a code review, as soon as it's approved I'll merge it.
Sorry for the long wait.

File: arduino-ide-extension/src/browser/contributions/board-selection.ts [hunk 1]
```diff
@@ -279,27 +278,27 @@ PID: ${PID}`;
 
       // First we show addresses with recognized boards connected,
       // then all the rest.
-      const sortedAddresses = Object.keys(ports);
-      sortedAddresses.sort((left: string, right: string): number => {
+      const sortedIDs = Object.keys(ports);
+      sortedIDs.sort((left: string, right: string): number => {
```
Comment (fstasi) [inline arduino-ide-extension/src/browser/contributions/board-selection.ts:282]:
```suggestion
      const sortedIDs = Object.keys(ports).sort((left: string, right: string): number => {
```

**Pull request merged**